jwt.go

```diff
@@ -189,20 +194,43 @@ func New(_ context.Context, next http.Handler, config *Config, _ string) (http.H
 			return nil, err
 		}
 		if len(jwtPlugin.jwkEndpoints) > 0 {
+			if config.ForceRefreshKeys {
+				jwtPlugin.forceRefreshCmd = make(chan chan<- struct{})
+			}
 			go jwtPlugin.BackgroundRefresh()
 		}
 	}
 	return jwtPlugin, nil
 }
 
 func (jwtPlugin *JwtPlugin) BackgroundRefresh() {
```

Comment:
Can you plz explain how exactly this function works now. I don't understand golang code that well.

Reply:
Sure. When this `ForceRefreshKeys` is set true, the channel `forceRefreshCmd` will be created for _**main routine**_ to send "Force Refresh" command to _**background refresh routine**_. The message of this channel (`make(chan [CHANNEL_MESSAGE_TYPE])`) is a one-time sending-only channel (`chan<-`) of an empty struct (`struct{}`) used as callback from the _**background refresh routine**_.

So for each of the two routines, the procedures are like:

_**Background Refresh Routine**_:
1. Listen to the channel `forceRefreshCmd` (in addition to a 15-min-timeout channel)
2. Receive a message (which is a one-time callback channel) from the channel `forceRefreshCmd`, indicating keys should be refreshed immediately
3. Refresh keys
4. Send an empty struct to the callback channel, in order to notify _**main routine**_ that keys are already refreshed

_**Main Routine**_ (when key is not found):
1. Create a new callback channel
2. Send the callback channel as a message to channel `forceRefreshCmd`, in order to notify that _**background refresh routine**_ should be refreshed immediately
3. Listen to the callback channel created on step 1
4. Receive a message of an empty struct, indicating that keys have been refreshed just now
5. Retry finding key ID of current JWT token and proceed